orc/src/main/java/com/netflix/iceberg/orc/TypeConversion.java

```diff
@@ -0,0 +1,172 @@
+        Types.MapType map = (Types.MapType) type;
+        columnIds.add(map.keyId());
+        TypeDescription key = toOrc(map.keyType(), columnIds);
+        columnIds.add(map.valueId());
+        return TypeDescription.createMap(key,
+            toOrc(type.asMapType().valueType(), columnIds));
+      }
+      default:
+        throw new IllegalArgumentException("Unhandled type " + type.typeId());
+    }
+  }
+
+  /**
+   * Convert an ORC schema to an Iceberg schema.
+   * @param schema the ORC schema
+   * @param columnIds the column ids
+   * @return the Iceberg schema
+   */
+  public Schema fromOrc(TypeDescription schema, int[] columnIds) {
+    return new Schema(convertOrcToType(schema, columnIds).asStructType().fields());
+  }
+
+  Type convertOrcToType(TypeDescription schema, int[] columnIds) {
+    switch (schema.getCategory()) {
+      case BOOLEAN:
+        return Types.BooleanType.get();
+      case BYTE:
+      case SHORT:
+      case INT:
+        return Types.IntegerType.get();
+      case LONG:
+        return Types.LongType.get();
+      case FLOAT:
+        return Types.FloatType.get();
+      case DOUBLE:
+        return Types.DoubleType.get();
+      case STRING:
+      case CHAR:
+      case VARCHAR:
+        return Types.StringType.get();
+      case BINARY:
+        return Types.BinaryType.get();
+      case DATE:
+        return Types.DateType.get();
+      case TIMESTAMP:
+        return Types.TimestampType.withoutZone();
+      case DECIMAL:
+        return Types.DecimalType.of(schema.getPrecision(), schema.getScale());
+      case STRUCT: {
+        List<String> fieldNames = schema.getFieldNames();
+        List<TypeDescription> fieldTypes = schema.getChildren();
+        List<Types.NestedField> fields = new ArrayList<>(fieldNames.size());
+        for(int c=0; c < fieldNames.size(); ++c) {
+          String name = fieldNames.get(c);
+          TypeDescription type = fieldTypes.get(c);
+          fields.add(Types.NestedField.optional(columnIds[type.getId()], name,
+              convertOrcToType(type, columnIds)));
+        }
+        return Types.StructType.of(fields);
+      }
+      case LIST: {
+        TypeDescription child = schema.getChildren().get(0);
+        return Types.ListType.ofOptional(columnIds[child.getId()],
+            convertOrcToType(child, columnIds));
+      }
+      case MAP: {
+        TypeDescription key = schema.getChildren().get(0);
+        TypeDescription value = schema.getChildren().get(1);
+        switch (key.getCategory()) {
+          case STRING:
+          case CHAR:
+          case VARCHAR:
+            return Types.MapType.ofOptional(columnIds[key.getId()],
+                columnIds[value.getId()], convertOrcToType(value, columnIds));
+          default:
+            throw new IllegalArgumentException("Can't handle maps with " + key +
+                " as key.");
+        }
+      }
+      default:
+        // We don't have an answer for union types -> blech.
```

Comment:
I'm interested to hear how ORC handles unions when only some of the columns are projected. This issue is why we never quite standardized unions in Parquet and why I'm reluctant to add them. (That, and I don't see a very distinct use case for them.)

I'm not against it in principle, but I'm skeptical and want to keep everything as small as possible until we're sure they are really necessary. Plus, many engines have no support for unions (e.g., Spark) and might not intend to support them.

Reply:
In practice, I haven't seen that case come up.

If you dropped some of the union children, those values would become null.

Reply:
I agree that returning null is the reasonable option, but doesn't that defeat the purpose of the guarantee that one branch of the union is non-null?

Reply:
Here's the discussion about a UNION type in Parquet, for reference: https://github.com/apache/parquet-format/pull/44

I think Alex makes some good points, like that projection should not affect the result, only the efficiency of the query.

Reply:
In ORC it is a bit easier. As I wrote elsewhere, the user can't have required fields, so if the user projected a single one of the union children, they would get nulls in the values where the value was one of the other children. This is the same as what would happen if the file had an extra child in the union. Those values would become null.

Reply:
One additional piece of the ORC format is that we don't push the metadata down into the leaf types. So the data for the union is a selector that says which child is selected for that value. Then that child's value is used. So if child 3 is picked, but child 3 was not in the projection, it would just have a null value. But the reader could tell that it was a child 3 value that was null.

https://orc.apache.org/api/hive-storage-api/org/apache/hadoop/hive/ql/exec/vector/UnionColumnVector.html